code/modules/power/chem_combustion.dm

```diff
@@ -0,0 +1,394 @@
+			if (current_reagent)
+				switch(reagent_id)
+					if ("foof", "kerosene", "nitrotri_dry", "dbreath")
+						average += (FUEL_QUALITY_VERY_HIGH * current_reagent.volume)
+
+					if ("blackpowder", "phlogiston", "napalm_goo", "sorium", "firedust")
+						average += (FUEL_QUALITY_HIGH * current_reagent.volume)
+
+					if ("fuel", "acetone")
+						average += (FUEL_QUALITY_AVERAGE * current_reagent.volume)
+
+					if ("oil", "butter", "diethylamine", "ethanol")
+						average += (FUEL_QUALITY_LOW * current_reagent.volume)
+
+					if ("plasma", "magnesium", "phosphorus")
+						average += (FUEL_QUALITY_VERY_LOW * current_reagent.volume)
+
+					else
+						var/datum/reagent/fooddrink/alcoholic/current_alcoholic_reagent = current_reagent
+						if (istype(current_alcoholic_reagent) && current_alcoholic_reagent.alch_strength >= 0.5)
+							average += 3
+
+		if (!average)
+			return 0
+
+		return average / R.total_volume
+
+#undef FUEL_QUALITY_VERY_HIGH
+#undef FUEL_QUALITY_HIGH
+#undef FUEL_QUALITY_AVERAGE
+#undef FUEL_QUALITY_LOW
+#undef FUEL_QUALITY_VERY_LOW
+
+	proc/start_engine()
+		if (!src.active)
+			if (!src.ready_to_start())
+				return 1
+
+			src.active = 1
+			src.UpdateIcon()
+			src.updateDialog()
+
+			if (!ON_COOLDOWN(src, "tractor", 2 SECOND))
+				playsound(src.loc, "sound/machines/tractorrev.ogg", 40, pitch=2)
+
+			return 0
+
+	proc/stop_engine()
+		if (src.active)
+			src.active = 0
+			src.UpdateIcon()
+			src.updateDialog()
+
+			if (!ON_COOLDOWN(src, "tractor", 2 SECOND))
+				playsound(src.loc, "sound/machines/tractorrev.ogg", 40, pitch=2)
+
+			return 0
+
+	proc/ready_to_start()
+		if (!anchored || !src.fuel_tank)
+			return 0
+
+		if (!src.get_average_volatility(src.fuel_tank.reagents) || !src.check_available_oxygen())
+			return 0
+
+		return 1
+
+	// Returns the concentration of oxygen in the available gas_mixture
+	proc/check_available_oxygen()
+		if (src.inlet_tank)
+			return src.check_tank_oxygen(src.inlet_tank)
+
+		var/turf/simulated/T = get_turf(src)
+		if (!istype(T))
+			return 0
+
+		if (!T.air || T.air.oxygen <= 0)
+			return 0
+
+		return T.air.oxygen / TOTAL_MOLES(T.air)
+
+	proc/check_tank_oxygen(obj/item/tank/T)
+		if (!src || !T || !T.air_contents)
+			return 0
+
+		if (T.air_contents.oxygen <= 0)
+			return 0
+
+		return T.air_contents.oxygen / TOTAL_MOLES(T.air_contents)
+
+	proc/get_output_cable()
+		var/list/cables = src.get_connections()
+
+		if (!cables.len)
+			return
+
+		for (var/obj/cable/C in cables)
+			if (C.get_powernet())
+				return C
+
+		return
+
+	proc/eject_fuel_tank(var/mob/user)
+		if (!src || !src.fuel_tank)
+			return
+
+		if (src.active)
+			src.visible_message("<span class='notice'>The [src] stops as the fuel tank is removed.</span>")
+			src.stop_engine()
+
+		src.fuel_tank.set_loc(get_turf(src))
+		if (istype(user))
+			user.put_in_hand_or_eject(src.fuel_tank)
+
+		src.fuel_tank = null
+		src.UpdateIcon()
+		playsound(src.loc, "sound/items/Deconstruct.ogg", 50, 1)
```

Comment:
Sound paths should use single quotes instead of double, for compile time checking of file stuff

Reply:
Fixed